dataset/batch_image.py

```diff
@@ -1,525 +1,767 @@
+        """
+        if isinstance(origin, str):
+            if origin == 'top_left':
+                origin = 0, 0
+            elif origin == 'center':
+                origin = np.maximum(0, np.asarray(background_shape) - image_shape) // 2
+            elif origin == 'random':
+                origin = (np.random.randint(background_shape[0]-image_shape[0]+1),
+                          np.random.randint(background_shape[1]-image_shape[1]+1))
+        return np.asarray(origin, dtype=np.int)
 
-    def _rotate_one(self, ix, components='images', angle=0, preserve_shape=True, **kwargs):
-        """ Rotate one image """
-        image = self.get(ix, components)
-        kwargs['reshape'] = not preserve_shape
-        new_image = scipy.ndimage.interpolation.rotate(image, angle, **kwargs)
-        return new_image
+    def _scale_(self, image, factor, preserve_shape=False, origin='center'):
+        """ Scale the content of each image in the batch.
 
-    @staticmethod
-    def _calc_origin(image, origin, shape):
-        if origin is None or origin == CROP_00:
-            origin = 0, 0
-        elif origin == CROP_CENTER:
-            origin_x = (image.shape[1] - shape[0]) // 2 if image.shape[1] > shape[0] else 0
-            origin_y = (image.shape[0] - shape[1]) // 2 if image.shape[0] > shape[1] else 0
-            origin = origin_x, origin_y
-        return origin
-
-    def _crop(self, components='images', origin=None, shape=None):
-        """ Crop all images in the batch """
-        if origin is not None or shape is not None:
-            images = self.get(None, components)
-
-            origin = self._calc_origin(images[0], origin, shape)
-            if shape is None:
-                shape = images.shape[2], images.shape[1]
-            if np.all(np.array(origin) + np.array(shape) > np.array(images.shape[1:3])):
-                shape = images.shape[2] - origin[0], images.shape[1] - origin[1]
-
-            x = slice(origin[0], origin[0] + shape[0])
-            y = slice(origin[1], origin[1] + shape[1])
-            new_images = images[:, y, x].copy()
-
-            setattr(self, components, new_images)
-
-    def _crop_image(self, image, origin, shape):
-        origin = self._calc_origin(image, origin, shape)
-        new_image = image[origin[1]:origin[1] + shape[1], origin[0]:origin[1] + shape[0]].copy()
-        return new_image
+        Resulting shape is obtained as original_shape * factor.
 
-    def _random_crop(self, components='images', shape=None):
-        if shape is not None:
-            images = self.get(None, components)
-            new_images = random_crop_numba(images, shape)
-            setattr(self, components, new_images)
-        return self
+        Parameters
+        -----------
+        components : str
+            component to get an image from
+        p : float
+            probability of applying the transforms
+        image : np.ndarray
+            image to scale
```

Comment:
This docstring is for `scale` method which does not have `image` parameter.

Reply:
+